### ISSUE TITLE
Fuel Gauge is relative to last dock, not full tank indicator

### ISSUE THREAD

**Issue by nabaco** [vegastrike/Vega-Strike-Engine-Source#105]
It seems that no matter how much I use overdrive, I don't see any fuel being used.

**Comment by royfalk** [comment]
First of, I'm not sure how much fuel ordinary drive consumes. It possible this is 0.
You can test by setting out in a direction and going to bed.
Second, it depends on the ship. The startup ship, llama has a pretty large fuel tank. If you switch to Gawain or one of the other small/light interceptors, you'll start to see fuel usage.
Again, I suspect that when we refactor, this issue will either go away or we'll know for sure how much fuel ordinary flight, overdrive, SPEC, jumps and firing guns take.

**Comment by stephengtuggy** [comment]
I can confirm that even the Llama definitely uses fuel, given enough time and engine burns. I think this issue can be closed.

**Comment by evertvorster** [comment]
It does use fuel... Just load it up with something heavy. 
However, after every docking, the fuel display will read 100% As in, you have 100% of the fuel you have taken off with, not 100% of the fuel tank. <- That's a bug, IMHO. (Different from this one)

**Comment by stephengtuggy** [comment]
I think this is fixed in 0.10.x, if I'm not mistaken.

**Comment by evertvorster** [comment]
Yes, apparently your ship gets refueled automatically on every landing. 
However, there is some disconnect between the title and the contents of this report. 

Fuel usage is something that needs further refinement. 

It should be a fairly straightforward calculation... A certain amount of fuel will produce a certain amount of thrust. 

mass x acceleration = thrust. 

Once we know how much thrust is needed to produce the change in velocity, we know the rate at which the fuel is being depleted, at least as far as motion is concerned. 

When a space ship is traveling at a constant speed, it is not changing its velocity, and so should not burn any fuel for that. However there should be some base level of fuel usage to keep life support and shields going. 
Charging capacitors and weapons usage should also deplete fuel at a faster rate. 
This can all be tied to how much energy they use, and how much energy the generator produces for a certain amount of fuel. 

However, changing the fuel use from a constant rate to something that is bound to what you are doing with your ship  constitutes quite a departure from how the game works now, and some code needs to be added to do this. Not sure if that is a priority right now. 

Of course this can be discussed so we know all the details before any coding starts.

**Comment by stephengtuggy** [comment]
@evertvorster yes, that type of fuel-consumption logic will need to be added. The thing is, though, that these types of calculations used to be present. They were removed at one point, in favor of a simpler formula. I think that was a mistake. Those calculations, or something similar, need to be added back.

**Comment by royfalk** [comment]
> It should be a fairly straightforward calculation... A certain amount of fuel will produce a certain amount of thrust.
> mass x acceleration = thrust.

That's not necessarily how the drives will work in the future. I don't want the engine to make assumptions like that.

> When a space ship is traveling at a constant speed, it is not changing its velocity, and so should not burn any fuel for that. 

Correct for current physics.

> However there should be some base level of fuel usage to keep life support and shields going.

That's already in the game. The reactor uses fuel to charge the capacitors and everything from guns to life support pull from there.

>> However, changing the fuel use from a constant rate to something that is bound to what you are doing with your ship constitutes quite a departure from how the game works now, and some code needs to be added to do this. Not sure if that is a priority right now.

> @evertvorster yes, that type of fuel-consumption logic will need to be added. The thing is, though, that these types of calculations used to be present. They were removed at one point, in favor of a simpler formula. I think that was a mistake. Those calculations, or something similar, need to be added back.

**A brief explanation**
I trimmed some code in order to refactor drive and afterburner into components. The old game made assumptions about carrying lithium(?) and ejecting it to space to move and while that may be true for VS, it's not true for WC for example.

A proper way to implement would be to add costs for not just drive and afterburner but every relevant `movable` action. Then, if you accelerate or decelerate, the game would call drive and it would subtract fuel accordingly. You could turn `realistic_fuel_mode` in settings.

I'm not very good at the physics of it, so someone would need to design the formulas for me to implement.

**Comment by evertvorster** [comment]
Oooh, I am a bit of a science fiction nut, so this discussion is right up my alley. 

> > It should be a fairly straightforward calculation... A certain amount of fuel will produce a certain amount of thrust.
> > mass x acceleration = thrust.
> 
> That's not necessarily how the drives will work in the future. I don't want the engine to make assumptions like that.
> 
Agreed that there might be different ways of propulsion in the future, some things that we can't even imagine right now. In Fiction there are wormholes and hyperspace etc, etc. I'm assuming that is covered by the jump points. 

For moving a ship around in normal space, you are bound by Newtonian laws. Unless humanity figures out something to push against in space, you are left with reaction engines. Even if we are somehow able to push against empty space, it will require energy to do so. 

Looking at the ships that are in the game right now, the exhaust appears to be ionic, which means a reaction engine. One way to make reaction engines more efficient is to push harder against the matter that is being thrown away. Since it takes an infinite amount of energy to accelerate anything to the speed of light, it follows that you could get infinite thrust by pushing hard enough on a single atom... So future engines may be a lot more efficient than the ones we have now. Which would lead to us having a tunable factor in how much thrust is provided by a certain amount of energy. 

Be that as it may, VS tries to implement physics as we understand them today, while bending the rules enough to make the game fun to play. 

Assuming that it costs fuel to provide thrust is a safe bet. How much fuel is used should be a tune-able factor until we have a value that "feels" right. 

Which brings me to the topic of the afterburner...
Of course, on terrestrial aircraft the afterburner is just dumping fuel into the exhaust to further accelerate the gasses to provide more thrust to counter the resistance of the air that the aircraft is pushing through. It has a nice natural negative feedback loop as you have more speed you also have more air resistance which limits the top speed of aircraft. 

In Space, there is no air resistance, and theoretically you can just continue to accelerate to any speed you want, because this negative feedback loop does not exist. 

The whole concept of maneuver mode is to make the game more fun to play. So, the only way to implement that is to have a ships computer limit the maximum speed of the ship to some value that would allow the ship to maneuver "like an aircraft". Once you turn the ship into "travel" mode it acts more like a space ship and you unlock the full potential of your thrusters at the cost of not being able to turn on a dime. 

So, the concept of an afterburner would just be an add-on that allows more energy into the engines to allow for a greater thrust.
This, in turn would give a faster acceleration, and so should just be a modifier to the standard gravities of thrust quoted in the space ship's specs. This should of course be on all the thrusters on the ship, as it increases the top speed that the computer feels is handle-able in manuever mode, and therefore it allows a higher "top speed" before regulating the speed. 

While changing course with your afterburner on, it will use more fuel. 

> > When a space ship is traveling at a constant speed, it is not changing its velocity, and so should not burn any fuel for that.
> 
> Correct for current physics.
> 
And current physics is what the player would expect. 

What this means for fuel usage is that once the ship is traveling at a constant speed, it is not consuming energy for acceleration, so even with the "afterburner" on, it would not be consuming any fuel on the thrusters. 


> > However there should be some base level of fuel usage to keep life support and shields going.
> 
> That's already in the game. The reactor uses fuel to charge the capacitors and everything from guns to life support pull from there.
> 
I suppose SPEC also uses energy and therefore fuel. 

> > > However, changing the fuel use from a constant rate to something that is bound to what you are doing with your ship constitutes quite a departure from how the game works now, and some code needs to be added to do this. Not sure if that is a priority right now.
> 
> > [@evertvorster](https://github.com/evertvorster) yes, that type of fuel-consumption logic will need to be added. The thing is, though, that these types of calculations used to be present. They were removed at one point, in favor of a simpler formula. I think that was a mistake. Those calculations, or something similar, need to be added back.
> 
Agreed, but this is something that should be added to a todo list somewhere. 

> **A brief explanation** I trimmed some code in order to refactor drive and afterburner into components. The old game made assumptions about carrying lithium(?) and ejecting it to space to move and while that may be true for VS, it's not true for WC for example.
>
Yeah, WC was not as concerned about realism and physics as VS. Agreed on having the engine support both models. 
 
> A proper way to implement would be to add costs for not just drive and afterburner but every relevant `movable` action. Then, if you accelerate or decelerate, the game would call drive and it would subtract fuel accordingly. You could turn `realistic_fuel_mode` in settings.
> 
Agreed. 
> I'm not very good at the physics of it, so someone would need to design the formulas for me to implement.
I would be happy to provide some outline of what works for the physics, and point to where the tunable factors should be to make this so that the game can be tuned to provide a pleasant user experience. 

We can then discuss what makes sense from a programming point of view. 
I believe that most, if not all, the elements are in the game engine already. It all needs connecting somehow. 

Do we move this discussion to its own thread for further consideration and reflection?

**Comment by stephengtuggy** [comment]
Yeah, probably should move this to its own thread.

Please note also that I just made the energy source for each component configurable in `configuration()` a few PRs ago. My default choices may not all be correct. It would be good to have someone review them.

**Comment by BenjamenMeyer** [comment]
just 2 cents on the issue:

components should have a configurable fuel usage that is dynamic based on their type essentially falling into one of three categories:
- exponential (x^2) - a `u` shaped curve
- linear (x=y) - a nice even line
- logarithmic (log(x)) - a nice `r` shaped curve

Light speed under Newtonian and Einsteinian Physics under current technology levels is on an exponential curve.

Low (cheap) engines should probably be exponential fuel cost; and as we move up (I to II to III to IV) we should probably move to linear cost and later (XX) to a logarithmic cost related to the technological advancements.

Lower tech races probably would be limited to exponential and linear costs.
High tech races would probably go across all three or even used a fixed cost.

**Comment by royfalk** [comment]
I'm pretty sure this specific bug is a combination of two things:
1. Llama.begin has no afterburner.
2. Llama.begin has a lot of fuel

Flying a short range fighter like Gawain makes fuel consumption much more visible and painful.
Having said that, we've also refactored fuel consumption and Llama no longer has near infinite fuel.

I've taken a look at the whole thread and it seems like there's a lot of important discussion there.
However, the bug itself is fixed.
@BenjamenMeyer @stephengtuggy what do you want to do here?

**Comment by BenjamenMeyer** [comment]
@royfalk if the bug itself is fixed then we can close this. Do we need to start a discussion or spin off another issue for the other for anything else? That said, we may want to capture some stuff in documentation too.
Thoughts?

**Comment by royfalk** [comment]
Opened #1520 to discuss and track this. Closing bug as it appears fixed.
